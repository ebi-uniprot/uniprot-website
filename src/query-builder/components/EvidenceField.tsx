import React, { useEffect, useState } from 'react';

import initializer from '../utils/fieldInitializer';

import { QueryBit, SearchTermType } from '../types/searchTypes';

const EvidenceField: React.FC<{
  field: SearchTermType;
  handleChange: (queryBit: QueryBit) => void;
  initialValue?: QueryBit;
}> = ({ field, handleChange, initialValue }) => {
  const [value, setValue] = useState(
    () => initializer(field, initialValue) as string
  );

  useEffect(() => {
    const trimmed = value.trim();
    if (trimmed) {
      handleChange({ [field.term]: trimmed });
    }
  }, [field, value, handleChange]);

  return (
    <label htmlFor="evidence_select">
      Evidence
      <select
        id="evidence_select"
        data-testid="evidence-select"
        value={value}
        onChange={(e) => setValue(e.target.value)}
      >
        {field.evidenceGroups?.map((group) => (
          <optgroup label={group.groupName} key={group.groupName}>
            {group.items.map((item: { code: string; name: string }) => (
              <option value={item.code} key={item.code}>
                {item.name}
              </option>
            ))}
          </optgroup>
        ))}
      </select>
    </label>
  );
};

export default EvidenceField;
