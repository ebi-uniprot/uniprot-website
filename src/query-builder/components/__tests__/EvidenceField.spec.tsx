import React from 'react';
import { render, fireEvent, getByTestId } from '@testing-library/react';
import EvidenceField from '../EvidenceField';

let rendered;
let props;

describe('EvidenceField component', () => {
  beforeEach(() => {
    props = {
      handleChange: jest.fn(),
      field: {
        id: 'ccev_webresource',
        itemType: 'single',
        term: 'ccev_webresource',
        dataType: 'string',
        fieldType: 'evidence',
        example: 'manual',
        evidenceGroups: [
          {
            groupName: 'foo',
            items: [
              {
                code: 'bar_code',
                name: 'bar',
              },
              {
                code: 'baz_code',
                name: 'baz',
              },
            ],
          },
        ],
      },
      initialValue: { ccev_webresource: 'baz_code' },
    };

    rendered = render(<EvidenceField {...props} />);
  });

  test('should change evidence', () => {
    const { getByTestId } = rendered;
    const evidenceSelect = getByTestId('evidence-select');
    fireEvent.change(evidenceSelect, {
      target: { value: 0 },
    });
    expect(props.handleChange).toBeCalledWith({
      ccev_webresource: 'baz_code',
    });
  });

  test('should render', () => {
    const { asFragment } = rendered;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should initialise', () => {
    const { getByTestId } = rendered;
    expect(getByTestId('evidence-select').value).toBe('baz_code');
  });
});
