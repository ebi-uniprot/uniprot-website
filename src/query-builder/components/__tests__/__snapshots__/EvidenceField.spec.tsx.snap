// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvidenceField component should render 1`] = `
<DocumentFragment>
  <label
    for="evidence_select"
  >
    Evidence
    <select
      data-testid="evidence-select"
      id="evidence_select"
    >
      <optgroup
        label="foo"
      >
        <option
          value="bar_code"
        >
          bar
        </option>
        <option
          value="baz_code"
        >
          baz
        </option>
      </optgroup>
    </select>
  </label>
</DocumentFragment>
`;
